src/internal/runnable/engine_factory.cpp

```diff
@@ -61,7 +63,7 @@ class FiberEngineFactory : public ::srf::runnable::EngineFactory
     }
 
   private:
-    virtual std::vector<std::shared_ptr<core::FiberTaskQueue>> get_next_n_queues(std::size_t count) = 0;
+    virtual std::vector<std::reference_wrapper<core::FiberTaskQueue>> get_next_n_queues(std::size_t count) = 0;
```

Comment:
I understand that you dont want to use a `shared_ptr` here to avoid a deadlock or circular references, but would a `weak_ptr` work here as well?

Whatever caused the circular reference due to lambda capture is capturing it because it needs it during destruction. Using a reference here cleans up ownership but doesnt guarantee that the object stays alive while its being used. A `weak_ptr` wouldnt influence the lifetime of the object similar to a reference, but you could make the guarantee that the object stayed alive while you were using it.

Reply:
I've restructured the code to make strong guarantees on the order creation and destruction for all resources. 

This will allow for more structured concurrency as we move toward coroutines. 

The only requirement is the Executor object is [uniquely] owned by a thread that is not constructed as part of the SRF runtime. 

Remember, you can't use a weak_ptr directly. In order to use a weak_ptr, you need to construct a shared_ptr via the lock method.

Reply:
> Remember, you can't use a weak_ptr directly. In order to use a weak_ptr, you need to construct a shared_ptr via the lock method.

Thats exactly what I am saying. If you have some code like this:
```cpp
// Get a single queue. next_queue will be of type std::reference_wrapper<core::FiberTaskQueue>
auto next_queue = this->get_next_n_queues(1)[0];

// Do some action with the queue.
next_queue.affinity();

// Sleep the fiber, when it wakes up next_queue could be destroyed
boost::this_fiber::sleep(1000);

// Try to perform some action on the reference again
next_queue.affinity(); // Crash
```
There is no guarantee that while you are holding onto `next_queue` it wont get destructed. It would have before since it was a `shared_ptr`.

Now if you used weak_ptr:
```cpp
// Get a single queue. next_queue will be of type std::weak_ptr<core::FiberTaskQueue>
auto next_queue = this->get_next_n_queues(1)[0];

// Need to convert my weak_ptr to a shared_ptr
std::shared_ptr<core::FiberTaskQueue> next_queue_ptr = next_queue.lock();

// Do some action with the queue.
next_queue_ptr->affinity();

// Sleep the fiber, when it wakes up next_queue_ptr is guaranteed to be alive
boost::this_fiber::sleep(1000);

// Try to perform some action on the reference again
next_queue_ptr->affinity();

// next_queue_ptr goes out of scope and can be destroyed
```

Reply:
There are strong guarantees that the FiberTaskQueues outlive all pipeline/segment/node/runnables.

Holding a shared pointer to a FiberTaskQueue which owns `system::Thread` on the thread/fiber is the culprit of the deadlock.

Reply:
Ok. Will just have to trust you on this one.